### ISSUE TITLE
rfc: should terser strip unused parameters during dead code elimination

### ISSUE THREAD

**Issue by cdaringe** [terser/terser#1080]
**Bug report or Feature request?**

Feature request.

In some cases, it can be discerned that parameters values can be erased.

Considerations working against this feature:

- `arguments` keyword
  - `var f = function () { doSomethingElse(arguments) }; f(() => "UNSAFE_TO_ERASE_THIS_LAMBA") `
- IIFEs mutating a variable, thus mutating a function implementation, thus making erasure dangerous
  - `var f = function () { doSomethingElse(arguments) }; f((() => { f = something_else; return () => ""; })()) `

However, in cases where 
  - the callee is protected from mutation (a const or module symbol)
  - the callee is a fat arrow expression (no `arguments` usage)
  - the callee does not use params

Then unused params may be _stripped_ from callers invoking the callee.

**Version (complete output of `terser -V` or specific git commit)**

`terser 5.9.0`

**Complete CLI command or `minify()` options used**

`terser --compress --mangle --ecma=2016 --module -- input.js`

**`terser` input**

```js
const f = () => null;
f(() => "ERASEME");
```

**`terser` output or error**

```js
const f=()=>null;f((()=>"ERASEME"));
```

**Expected result**
```js
const f=()=>null;f();
```

**Comment by TimvdLippe** [comment]
I am not a maintainer of Terser, but as a user I would expect such a feature from a type checker, not per se a minifier. I would expect the type checker to complain that the number of arguments is wrong and that you should supply fewer arguments. Would adopting a type checker help in your situation?

**Comment by jridgewell** [comment]
I agree with @TimvdLippe here. If this is source code you've written, it's a bit strange. Is it possible this is the output of some other build step (in which case, supporting it becomes more of a priority).

**Comment by cdaringe** [comment]
This is a common pattern when supplying different implementations between build target environments.

process.env.NODE_ENV === “production” ? implNotUsingParams : implUsingParams

Agreed that it is funky to write such an impl directly, but it’s an optimization that is realistically practical in JS projects. With Webpack define plugin, Babel/TS/swc transforms, all generating code eventually piped into terser, it’s not unrealistic to expect this case

**Comment by fabiosantoscode** [comment]
In the case where Terser can turn a function call into an IIFE (where it's provable that it's only called once), Terser already removes extraneous arguments (the call had 3 arguments before Terser):

![image](https://user-images.githubusercontent.com/1611595/135848286-363940e1-c4dc-49da-928e-dbd52ed9841d.png)

The tricky part is keeping track of how many times something was called. Specifically,

> process.env.NODE_ENV === “production” ? implNotUsingParams : implUsingParams

Would be pretty hard to do in a single compressor pass (first we need to remove that ternary).

That said, for this optimization specifically: How many bytes would this save? There's lots of things Terser doesn't do, because it's hard to maintain a lot of working features, and the project is already quite large.

**Comment by cdaringe** [comment]
Thanks for the great feedback.

1. How many bytes saved?

```ts
implNotUsingParams(function functionThatCouldBeErased () {
  // how many bytes are saved would be
  // equal to however many bytes terser
  // could compress this function down to.
  // I'm passing this CB to a function that won't use it
  // so the MAX_BYTES_SAVED can be quite high.
  // sizeOfFn.toString().length bytes :)
}, plusThis, ...andThese)
```

In the case that prompted this issue, pumping my fn implementation into a minifier, it was ~300-400 bytes. Could be a single byte, could be a megabyte!


2. On optimization passes

> The tricky part is keeping track of how many times something was called. Specifically,
> process.env.NODE_ENV === “production” ? implNotUsingParams : implUsingParams`

I used this as an example, but in actuality, I get into these scenarios with upstream build tools, such as webpack/etc. By the time this snippet makes it to terser, it is reduced to:

- `"production === "production" ? implNotUsingParams : implUsingParams`, or 
- `true  ? implNotUsingParams : implUsingParams` 

i'm not sure certain which one is more likely. 

This simplified input to terser may or may not reduce the need for a multi pass. I would opt in to pay the runtime cost for a multi pass reduction if it meant support for the feature. Speed matters 101%. Browser bytes is something I prioritize even higher than build speed.

> In the case where Terser can turn a function call into an IIFE

Thanks. Ya, this is a different case indeed. This fn could be called many times, but the inputs are _known to do nothing_, so we can save!

> There's lots of things Terser doesn't do, because it's hard to maintain a lot of working features,

Ya, I appreciate that. I additionally acknowledge that this could be argued as an edge case as well.

For me, the driving motivation is really supporting a development and production workflow in my JS projects. Were it not for this objective, I would not end up with the seemingly obscure code as @TimvdLippe & @jridgewell pointed out. To implement this feature is to implicitly support browser patterns where implementations are swapped based on build targets. That may not be a goal of this project, or even a consideration to tailor towards.

In development, we load different features into the application than we do at production time. Extra config, plugins, widgets, etc are loaded into our applications to enhance the developer experience. By the time we go to production, none of that is needed, and we desire to strip it out. For static resources, dead-code-elim is great. `isProd ? prodImpl : devImpl` strips out `devImpl`, and everyone is happy! However, it just gets a bit messier when your resources are composed with other resources, e.g. 

```ts
const a = resource => isProd ? prodImpl : devImpl(resource)  // devImpl dead-code-elim'd
const b = fn(resource) // resource param IS NOT dead-code-elim'd
```
 I can _definitely_ handle all this in userland code:

```ts
const a = resource => isProd ? prodImpl : devImpl(resource) // devImpl dead-code-elim'd
const b = isProd ? fn() : fn(resource) // resource param IS dead-code-elim'd
```

The downside to handling it in userland code (in fn `b`) as demonstrated is that _it's redundant_. I **already expressed my production optimization** in fn `a`, but now must force it onto my customer who is defining function `b`. Not a problem--the caller should probably be aware of browser impact. However, it's certainly _more_ desirable if I can handle that conditionality once and let the compiler work its optimization to handle the rest.

Food for thought. Please feel empowered to close the issue if we do not want to pursue further, otherwise I am happy to continue the dialoge.

**Comment by TomerAberbach** [comment]
A good use-case for this is performing DCE for patterns like: https://github.com/alexreardon/tiny-invariant

Right now they rely on Babel plugins like [babel-plugin-dev-expression](https://www.npmjs.com/package/babel-plugin-dev-expression) to get the code into the right structure for minifiers like Terser to optimize it. I think only the `process.env.NODE_ENV` expression replacement would be needed (without method-call/if-statement restructuring) if Terser stripped unused parameter arguments.

**Comment by fabiosantoscode** [comment]
It's not hard to keep track of which functions are called. I've done some experiments at the end of last year, to optimize the "options argument" pattern.

However across module boundaries it gets much, much more complicated. Other tools do the bundling and there's currently no way for them to tell Terser where each module starts (and where/how imports/exports are located and represented), especially given the diversity of plugins, bundlers and bundling options involved.

**Comment by TomerAberbach** [comment]
Ah gotcha, yeah I agree it would get complicated with modules!

I guess the case I had in mind was after bundling all the code (dependencies plus the application code) into a single file, in which case it seems tractable. But I guess maybe you run into the same problem you mentioned if code splitting is done. Is that right?

By the way, I looked through the code a bit and noticed Terser [already does the analysis described here](https://github.com/terser/terser/blob/master/lib/compress/index.js#L1631-L1670) for code like this:
```js
((a, b) =>  a + b)(1, 2, 3, 4);
```

It outputs:
```js
((a, b) =>  a + b)(1, 2);
```

So it seems that the work required here would be to make Terser analyze the lefthand side of the call more (e.g. identifier as well if it is resolvable to a function with known number of arguments).

**Comment by fabiosantoscode** [comment]
It appears that this was fixed in the meantime.

```js
{
  module: false,
  compress: {
    global_defs: {
      "process.env.NODE_ENV": "production"
    },
    passes: 2
  }
}
```

Input:
```js
const f = process.env.NODE_ENV === 'production' ? () => null : (a) => console.log(a());
f(() => "ERASEME");
```

Output: `const f=()=>null;`

It does need 2 passes, but I think that's reasonable.
